### PR TITLE
[2/N] improvements to saving and loading callback state

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #6467

Adds an upgrade path for all checkpoints that we load with either `Trainer(resume_from_checkpoint=)` or `LightningModule.load_from_checkpoint`. It will convert a checkpoint saved with an older version and make it compatible with the new Lightning version. 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -29,6 +29,14 @@ class Callback(abc.ABC):
     Subclass this class and override any of the relevant hooks
     """
 
+    @property
+    def state_identifier(self) -> str:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:33]: Or save the __target__ path directly ?

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:33]: Example: "pytorch_lightning.callbacks.ModelCheckpoint"

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 2]
```diff
@@ -86,3 +94,34 @@ def test_resume_legacy_checkpoints(tmpdir, pl_version: str):
...
+        default_root_dir=tmpdir,
+        max_steps=1,
+        callbacks=[callback],
+    )
+    trainer.fit(model)
+    # simulate old format where type(callback) was the key
+    new_checkpoint = torch.load(Path(tmpdir, "last.ckpt"))
+    old_checkpiont = deepcopy(new_checkpoint)
+    old_checkpiont["callbacks"] = {type(callback): new_checkpoint["callbacks"]["ModelCheckpoint"]}
+    torch.save(old_checkpiont, Path(tmpdir, "old.ckpt"))
```
Comment (tchaton) [inline tests/checkpointing/test_legacy_checkpoints.py:119]: Neat !

File: pytorch_lightning/utilities/migration/migrations.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# v1.3.1
+def migrate_callback_state_identifiers(checkpoint):
+    if "callbacks" not in checkpoint:
+        return
+    callbacks = checkpoint["callbacks"]
+    checkpoint["callbacks"] = dict((callback_type.__name__, state) for callback_type, state in callbacks.items())
+    return checkpoint
+
+
+def migrate_checkpoint(checkpoint: dict):
```
Comment (tchaton) [inline pytorch_lightning/utilities/migration/migrations.py:37]: Should we expose this in the documentation ?

File: pytorch_lightning/utilities/upgrade_checkpoint.py [hunk 4]
```diff
@@ -57,4 +34,7 @@ def upgrade_checkpoint(filepath):
 
     log.info("Creating a backup of the existing checkpoint file before overwriting in the upgrade process.")
     copyfile(args.file, args.file + ".bak")
-    upgrade_checkpoint(args.file)
+    with pl_legacy_patch():
+        checkpoint = torch.load(args.file)
+    migrate_checkpoint(checkpoint)
+    torch.save(checkpoint, args.file)
```
Comment (ananthsub) [inline pytorch_lightning/utilities/upgrade_checkpoint.py:40]: use fsspec here to enable remote files
